README.md

# CASSANDRA-LOADER

## Introduction

cassandra-loader is a general-purpose, delimited-file, bulk loader for 
Cassandra. It supports a number of configuration options to enable bulk 
loading of various types of delimited files, including
* comma-separated values
* tab-separated values
* customer delimiter-separated values
* header row
* comma as decimal separator
* ...

## Getting it

### Downloading

This utility has already been built, and is available at
https://github.com/brianmhess/cassandra-loader/releases/download/v0.0.27/cassandra-loader

Get it with wget:

```
wget https://github.com/brianmhess/cassandra-loader/releases/download/v0.0.27/cassandra-loader
```

### Building

To build this repository, simply clone this repo and run:

```
gradle loader
```

All of the dependencies are included (namely, the Java driver - currently
version 3.0.0).  The output will be the cassandra-loader executable
in the build directory.  There will also be an jar with all of the
dependencies included in the build/libs/cassandra-loader-uber-<version>.jar

## Documentation 

To extract this README document, simply run (on the cassandra-loader
executable - (e.g., on build/cassandra-loader):

```
jar xf cassandra-loader README.md
```

## Run

To run cassandra-loader, simply run the cassandra-loader executable 
(e.g., located at build/cassandra-loader):

```
cassandra-loader
```

If you built this with gradle, you can also run:

```
gradle run
```

This will print the usage statement.

The following will load the `myFileToLoad.csv` file into the Cassandra 
cluster at IP address 1.2.3.4 into the `test.ltest` column family where 
the myFileToLoad file has the format of 4 columns - and it gets the
data type information from the database - and using the default options:

```
cassandra-loader -f myFileToLoad.csv -host 1.2.3.4 -schema "test.ltest(a, b, c, d)"
```

## Options:

 Switch           | Option             | Default                    | Description
-----------------:|-------------------:|---------------------------:|:----------
 `-configFile`    | Filename           | none                       | Filename of configuration options 
 `-f`             | Filename           | &lt;REQUIRED&gt;                 | Filename to load - required.
 `-host`          | IP Address         | &lt;REQUIRED&gt;                 | Cassandra connection point - required.
 `-format`        | Input format       | delim                            | Format of the data.  Options are "delim" or "json".
 `-schema`        | CQL schema         |                                  | Schema of input data - required for delim In the format "keySpace.table(col1,col2,...)" and in the order that the data will be in the file.
 `-keyspace`      | Keyspace name      |                            | Name of keyspace (case sensitive) to load in to - required for json
 `-table`         | Table name         |                            | Name of table (case sensitive) to load in to - required for json
 `-port`          | Port Number        | 9042                       | Cassandra native protocol port number
 `-user`          | Username           | none                       | Cassandra username
 `-pw`            | Password           | none                       | Cassandra password
 `-ssl-truststore-path` | Truststore Path     | none                | Path to SSL truststore
 `-ssl-truststore-pw`  | Truststore Password | none                | Password to SSL truststore
 `-ssl-keystore-path`   | Keystore Path       | none                | Path to SSL keystore
 `-ssl-keystore-path`   | Keystore Password   | none                | Password to SSL keystore
 `-consistencyLevel`    | Consistency Level   | ONE                 | CQL Consistency Level
 `-numThreads`    | Number of threads  | Number of CPUs             | Number of threads to use (one per file)
 `-numFutures`    | Number of Futures  | 1000                       | Number of Java driver futures in flight.
 `-numRetries`    | Number of retries  | 1                          | Number of times to retry the INSERT before declaring defeat.
 `-queryTimeout`  | Timeout in seconds | 2                          | Amount of time to wait for a query to finish before timing out.
 `-ttl`           | Time To Live       | none                       | TTL to use when inserting these rows
 `-delim`         | Delimiter          | ,                          | Delimiter to use
 `-charsPerColumn`| Characters per column | 4096                    | Maximum characters per column
 `-nullString`    | Null String        | &lt;empty string&gt;             | String to represent NULL data
 `-boolStyle`     | Boolean Style      | TRUE_FALSE                 | String for boolean values.  Options are "1_0", "Y_N", "T_F", "YES_NO", "TRUE_FALSE".
 `-decimalDelim`  | Decimal delimiter  | .                          | Delimiter for decimal values.  Options are "." or ","
 `-dateFormat`    | Date Format String | default for Locale.ENGLISH | Date format string as specified in the SimpleDateFormat Java class: http://docs.oracle.com/javase/7/docs/api/java/text/SimpleDateFormat.html
 `-skipRows`      | Rows to skip       | 0                          | Number of rows to skip at the beginning of the file
 `-skipCols`      | Columns to skip    | &lt;not set&gt;                  | Comma-separated list of columns to skip loading (0-counted)
 `-maxRows`       | Max rows to read   | -1                         | Maximum rows to read (after optional skipping of rows).  -1 signifies all rows.
 `-maxErrors`     | Max parse errors   | 10                         | Maximum number of rows that do not parse to allow before exiting.
 `-maxInsertErrors`| Max insert errors | 10                         | Maximum number of rows that do not insert to allow before exiting.
 `-badDir`        | Bad directory      | current directory          | Directory to write badly parsed and badly inserted rows - as well as the log file.
 `-rate`          | Ingest rate        | unlimited                  | Maximum rate to insert data - in rows/sec.
 `-progressRate`  | Progress rate      | 100000                     | How often to report the ingest rate (number of rows)
 `-rateFile`      | Rate Stats File    | &lt;not set&gt;                  | File to contain CSV rate statistics
 `-successDir`    | Success directory  | &lt;not set&gt;                  | Location to move successfully loaded files
 `-failureDir`    | Failure directory  | &lt;not set&gt;                  | Location to move files that failed to load
 `-batchSize`     | Batch size         | 1                          | Size of unlogged batches. If set to 1 then no batching.
 `-comment`       | Comment character  | $lt;not set&gt;                  | Comment character.

## Comments

### Using stdin

You can send data in on stdin by specifying the filename (via the -f switch) as "stdin" (case insensitive).
That way, you could pipe data in from other commands:

```
grep IMPORTANT data.csv | cassandra-loader -f stdin -h 1.2.3.4 -schema "test.itest(a, b)"
```

### Support for collections 

Collections are supported.  Their format is the CQL native one.
Sets are started with '{' and ended with '}' and enclose a comma-separated list
{1,2,3} or {"a","b","c"}
Lists are started with '[' and ended with ']' and enclose a comma-separated list
[1,2,3] or ["a","b","c"]
Maps are started with '{' and ended with '}' and enclose a comma-separated list
of pairs that are separated by ':'
{1:1,2:2,3:3} or {"a":1, "b":2, "c":3}
All collections must be enclosed in double-quotes.

### Username/Password

If you specify either the username or the password, then you must specify both.

### Boolean Style

boolStyle is a case-insensitive test of the True and False strings.  For the
different styles, the True and False strings are as follows:

```
    Style   | True | False
------------|------|-------
     0_1    |    1 |     0 
     Y_N    |    Y |     N 
     T_F    |    T |     F 
   YES_NO   |  YES |    NO 
 TRUE_FALSE | TRUE | FALSE 
```

### Configuration file

configFile is a file with configuration options that are formatted just like on
the command line.  This allows you to not specify arguments on the command line.
For example, you can specify passwords in the configFile and avoid having them on
the command line.  The format is one switch and option per line:

```
-pw mypassword
-host 1.2.3.4
```

### Miscelaneous

numFutures is a way to control the level of parallelism, but at some point 
too many will actually slow down the load.  The default of 500 is a decent 
place to start.

If you use quoted strings, you need to use double-quotes.  To escape a double-quote inside a quoted string, use the backslash to escape it ("\"").  To create a backslash inside a quoted string, use two backslashes in a row ("\\").  If you quote your string, it will not be trimmed, but if you do not quote your string it will be trimmed.

Loading into counter columns is not supported.

The default nullString is the empty string.  If you want empty strings to be saved as empty strings, set the nullString to something else.

If you do not set the successDir then files that successfully loaded will remain in their input directory.  The same is true for failed files if you do not set the failureDir.  You cannot set either if the input file is "stdin".

When using `jsonline`, all JSON field names are case-sensitive.  When using `jsonline` or `jsonarray`, the `-keyspace` and `-table` arguments are case-sensitive.

## Usage Statement:

```
version: 0.0.27
Usage: -f <filename> -host <ipaddress> [OPTIONS]
OPTIONS:
  -schema <schema>                   Table schema (when using delim)
  -table <tableName>                 Table name (when using json)
  -keyspace <keyspaceName>           Keyspace name (when using json)
  -configFile <filename>             File with configuration options
  -delim <delimiter>                 Delimiter to use [,]
  -charsPerColumn <chars>            Max number of chars per column [4096]
  -dateFormat <dateFormatString>     Date format for TIMESTAMP [default for Locale.ENGLISH]
  -localDateFormat <formatString>    Date format for DATE [yyyy-MM-dd]
  -nullString <nullString>           String that signifies NULL [none]
  -comment <commentString>           Comment symbol to use [none]
  -skipRows <skipRows>               Number of rows to skip [0]
  -skipCols <columnsToSkip>          Comma-separated list of columsn to skip in the input file
  -maxRows <maxRows>                 Maximum number of rows to read (-1 means all) [-1]
  -maxErrors <maxErrors>             Maximum parse errors to endure [10]
  -badDir <badDirectory>             Directory for where to place badly parsed rows. [none]
  -port <portNumber>                 CQL Port Number [9042]
  -user <username>                   Cassandra username [none]
  -pw <password>                     Password for user [none]
  -ssl-truststore-path <path>        Path to SSL truststore [none]
  -ssl-truststore-pwd <pwd>           Password for SSL truststore [none]
  -ssl-keystore-path <path>          Path to SSL keystore [none]
  -ssl-keystore-pwd <pwd>             Password for SSL keystore [none]
  -consistencyLevel <CL>             Consistency level [LOCAL_ONE]
  -numFutures <numFutures>           Number of CQL futures to keep in flight [1000]
  -batchSize <batchSize>             Number of INSERTs to batch together [1]
  -decimalDelim <decimalDelim>       Decimal delimiter [.] Other option is ','
  -boolStyle <boolStyleString>       Style for booleans [TRUE_FALSE]
  -numThreads <numThreads>           Number of concurrent threads (files) to load [num cores]
  -queryTimeout <# seconds>          Query timeout (in seconds) [2]
  -numRetries <numRetries>           Number of times to retry the INSERT [1]
  -maxInsertErrors <# errors>        Maximum INSERT errors to endure [10]
  -rate <rows-per-second>            Maximum insert rate [50000]
  -progressRate <num txns>           How often to report the insert rate [100000]
  -rateFile <filename>               Where to print the rate statistics
  -successDir <dir>                  Directory where to move successfully loaded files
  -failureDir <dir>                  Directory where to move files that did not successfully load
  -nullsUnset [false|true]           Treat nulls as unset [faslse]
  -format [delim|jsonline|jsonarray] Format of data: delimited or JSON [delim]
  -table <tableName>                 Table name (when using JSON)
  -keyspace <keyspaceName>           Keyspace name (when using JSON)
  -ttl <TTL>                         TTL for all rows in this invocation [unset]


Examples:
cassandra-loader -f /path/to/file.csv -host localhost -schema "test.test3(a, b, c)"
cassandra-loader -f /path/to/directory -host 1.2.3.4 -schema "test.test3(a, b, c)" -delim "\t" -numThreads 10
cassandra-loader -f stdin -host localhost -schema "test.test3(a, b, c)" -user myuser -pw mypassword
```

##Examples:

Load file /path/to/file.csv into the test3 table in the test keyspace using
the cluster at localhost.  Use the default options:

```
cassandra-loader -f /path/to/file.csv -host localhost -schema "test.test3(a, b, c)"
```

Load all the files from /path/to/directory into the test3 table in the test
keyspace using the cluster at 1.2.3.4.  Use 10 threads and use tab as the
delimiter:

```
cassandra-loader -f /path/to/directory -host 1.2.3.4 -schema "test.test3(a, b, c)" -delim "\t" -numThreads 10
```

Load the data from stdin into the test3 table in the test keyspace using the
cluster at localhost.  Use "myuser" as the username and "mypassword" as the
password:

```
cassandra-loader -f stdin -host localhost -schema "test.test3(a, b, c)" -user myuser -pw mypassword
```

##Sample

Included here is a set of sample data.  It is in the sample/ directory.
You can set up the table and keyspace by running:

```
cqlsh -f sample/cassandra-schema.cql
```

To load the data, run:

```
cd sample
./load.sh
```

To check that things have succeeded, you can run:

```
wc -l titanic.csv
```

And:

```
cqlsh -e "SELECT COUNT(*) FROM titanic.surviors"
```

Both should return 891.



## cassandra-unloader
cassandra-unloader is a utility to dump the contents
of a Cassandra table to delimited file format.  It uses
the same sorts of options as cassandra-loader so that the
output of cassandra-unloader could be piped into 
cassandra-loader:

```
cassandra-unloader -f stdout -host host1 -schema "ks.table(a,b,c)" | cassandra-loader -f stdin -host host2 -schema "ks2.table2(x,y,z)"
```

Get it with wget:
```
wget https://github.com/brianmhess/cassandra-loader/releases/download/v0.0.27/cassandra-unloader
```

To build, run:

```
gradle unloader
```

To run cassandra-unloader, simply run the cassandra-unloader executable 
(e.g., located at build/cassandra-unloader):

```
cassandra-unloader
```

###Usage statement:

```
version: 0.0.27
Usage: -f <outputStem> -host <ipaddress> -schema <schema> [OPTIONS]
OPTIONS:
  -configFile <filename>             File with configuration options
  -format [delim|jsonline|jsonarray] Format of data: delimited or JSON [delim]
  -delim <delimiter>                 Delimiter to use [,]
  -dateFormat <dateFormatString>     Date format for TIMESTAMP [default for Locale.ENGLISH]
  -localDateFormat <FormatString>    Date format for DATE [yyyy-MM-dd]
  -nullString <nullString>           String that signifies NULL [none]
  -port <portNumber>                 CQL Port Number [9042]
  -user <username>                   Cassandra username [none]
  -pw <password>                     Password for user [none]
  -ssl-truststore-path <path>        Path to SSL truststore [none]
  -ssl-truststore-pwd <pwd>           Password for SSL truststore [none]
  -ssl-keystore-path <path>          Path to SSL keystore [none]
  -ssl-keystore-pwd <pwd>             Password for SSL keystore [none]
  -consistencyLevel <CL>             Consistency level [LOCAL_ONE]
  -decimalDelim <decimalDelim>       Decimal delimiter [.] Other option is ','
  -boolStyle <boolStyleString>       Style for booleans [TRUE_FALSE]
  -numThreads <numThreads>           Number of concurrent threads to unload [5]
  -beginToken <tokenString>          Begin token [none]
  -endToken <tokenString>            End token [none]
  -where <predicate>                 WHERE clause [none]
  -fetchSize <fetchSize>             Fetch size to use [0]
```

A few simple examples using the `-where` are as follows:

```
cassandra-unloader -host localhost -f stdout -schema "testks.testtable(pkey,ccol,x,y)" -where "pkey=5"
cassandra-unloader -host localhost -f stdout -schema "testks.testtable(pkey,ccol,x,y)" -where "x = 100 ALLOW FILTERING"
```


ChangeLog.md
## 0.0.27
- Added support for specifying a TTL (Issue 67)

## 0.0.26
- Fixed issue with long schemas (Issue 65)

## 0.0.25
- Added support for Fetch Size in cassandra-unloader

## 0.0.24
- Added support for DATE (and associated -localDateFormat option)
- Added support for SHORTINT and TINYINT
- Fixed an issue with special characters in column names (Issue 59)

## 0.0.23
- Fixed case of commas in Map keys
- Fixed keystore/truststore-pw typo (was pwd)
- Fixed quoting of non-collections
- Added support for supplying comment character

## 0.0.22
- Added support for gzipped input files
- Fixed issue with quoted values (and in collections)

## 0.0.21
- Added support for jsonarray (one JSON array per file)
- Added support for jsonline (one JSON per line)
- Changed JVM settings to 1GB heap
- Added -charsPerColumn (default to 4096) to enable Univocity optimizations
- Removed explicit queries to metadata tables and use driver API calls

## 0.0.20
- Fixed delimiter in MapParser
- Catch NULLs in Map/Set/List parsing and throw in BADPARSE

## 0.0.19
- Converted to parse with Univocity CSV parser
- Added -where

## 0.0.18
- Support for Cassandra 3.0

## 0.0.17
- Fixed null collection issue / NPE (Issue 8)

## 0.0.16
- Unloader will quote collections (which the loader expects)
- Fixed collection issue (Issue 14)
- BLOBs are now Base64 encoded on unload, and should be Base64 to load (Issue 15)
- Support for quoted keyspace, table, and column names

## 0.0.15
- Better error handling for case when C* inserts are failing

## 0.0.14
- Updated cassandra-unloader to add support for collections,
	consistency level, ssl, etc

## 0.0.13
- Added configFile
- added ssl options (with truststore and keystore)

## 0.0.12
- Added a rateFile to output CSV rate statistics
- added -skipCols to skip input columns

## 0.0.11
- Added support for quoted Keyspaces, Tables, and Columns

## 0.0.10
- You want collections?  You got 'em
- Added progress reporting - you can specify the rate at which
	the rate is reported via the -progressRate option
- Refactored RateLimiting - added it to a new RateLimitingSession
- Laid groundwork for Dynamic rate limiting - to be worked out
	once we find a way to collect the right statistic

## 0.0.9
- Added -successDir and -failureDir
- Added return codes for the loader and unloader
- Refactored BoolStyle
- Cleaned up the readme a bit



CqlDelimLoad.java
package hess.loader;

import hess.loader.parser.BooleanParser;
import hess.loader.parser.CqlDelimParser;

import java.lang.System;
import java.lang.String;
import java.lang.Integer;
import java.util.List;
import java.util.ArrayList;
import java.util.Map;
import java.util.HashMap;
import java.util.Set;
import java.util.HashSet;
import java.util.Deque;
import java.util.ArrayDeque;
import java.util.Locale;
import java.io.File;
import java.io.BufferedWriter;
import java.io.FileWriter;
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.text.ParseException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Callable;
import java.util.concurrent.Future;
import java.util.concurrent.ExecutionException;

import com.datastax.driver.core.Cluster;
import com.datastax.driver.core.Session;
import com.datastax.driver.core.PreparedStatement;
import com.datastax.driver.core.BoundStatement;
import com.datastax.driver.core.ResultSetFuture;
import com.datastax.driver.core.policies.TokenAwarePolicy;
import com.datastax.driver.core.policies.DCAwareRoundRobinPolicy;


public class CqlDelimLoad {
    private String host = null;
    private int port = 9042;
    private String username = null;
    private String password = null;
    private Cluster cluster = null;
    private Session session = null;
    private int numFutures = 1000;

    private String cqlSchema = null;

    private long maxErrors = 10;
    private long skipRows = 0;
    private long maxRows = -1;
    private String badDir = null;
    private BufferedWriter badWriter = null;
    private String filename = null;

    private Locale locale = null;
    private BooleanParser.BoolStyle boolStyle = null;
    private String dateFormatString = null;
    private String nullString = null;
    private String delimiter = null;
    private boolean delimiterInQuotes = false;

    private int numThreads = 5;

    private String usage() {
	String usage = "Usage: -f <filename> -host <ipaddress> -schema <schema> [OPTIONS]\n";
	usage = usage + "OPTIONS:\n";
	usage = usage + "  -delim <delimiter>             Delimiter to use [,]\n";
	usage = usage + "  -delmInQuotes true             Set to 'true' if delimiter can be inside quoted fields [false]\n";
	usage = usage + "  -dateFormat <dateFormatString> Date format [default for Locale.ENGLISH]\n";
	usage = usage + "  -nullString <nullString>       String that signifies NULL [none]\n";
	usage = usage + "  -skipRows <skipRows>           Number of rows to skip [0]\n";
	usage = usage + "  -maxRows <maxRows>             Maximum number of rows to read (-1 means all) [-1]\n";
	usage = usage + "  -maxErrors <maxErrors>         Maximum errors to endure [10]\n";
	usage = usage + "  -badDir <badDirectory>         Directory for where to place badly parsed rows. [none]\n";
	usage = usage + "  -port <portNumber>             CQL Port Number [9042]\n";
	usage = usage + "  -user <username>               Cassandra username [none]\n";
	usage = usage + "  -pw <password>                 Password for user [none]\n";
	usage = usage + "  -numFutures <numFutures>       Number of CQL futures to keep in flight [1000]\n";
	usage = usage + "  -decimalDelim <decimalDelim>   Decimal delimiter [.] Other option is ','\n";
	usage = usage + "  -boolStyle <boolStyleString>   Style for booleans [TRUE_FALSE]\n";
	usage = usage + "  -numThreads <numThreads>       Number of concurrent threads (files) to load\n";
	usage = usage + "\n\nExamples:\n";
	usage = usage + "cassandra-loader -f /path/to/file.csv -host localhost -schema \"test.test3(a int, b int, c int)\"\n";
	usage = usage + "cassandra-loader -f /path/to/directory -host 1.2.3.4 -schema \"test.test3(a int, b int, c int)\" -delim \"\\t\" -numThreads 10\n";
	usage = usage + "cassandra-loader -f stdin -host localhost -schema \"test.test3(a int, b int, c int)\" -user myuser -pw mypassword\n";
	return usage;
    }
    
    private boolean validateArgs() {
	if (0 >= numFutures) {
	    System.err.println("Number of futures must be positive");
	    return false;
	}
	if (0 > skipRows) {
	    System.err.println("Number of rows to skip must be non-negative");
	    return false;
	}
	if (0 >= maxRows) {
	    System.err.println("Maximum number of rows to load must be positive");
	    return false;
	}
	if (0 > maxErrors) {
	    System.err.println("Maximum number of errors must be non-negative");
	    return false;
	}
	if (numThreads < 1) {
	    System.err.println("Number of threads must be non-negative");
	    return false;
	}
	if (!filename.equalsIgnoreCase("stdin")) {
	    File infile = new File(filename);
	    if ((!infile.isFile()) && (!infile.isDirectory())) {
		System.err.println("The -f argument needs to be a file or a directory");
		return false;
	    }
	    if (infile.isDirectory()) {
		File[] infileList = infile.listFiles();
		if (infileList.length < 1) {
		    System.err.println("The directory supplied is empty");
		    return false;
		}
	    }
	}
	if ((null == username) && (null != password)) {
	    System.err.println("If you supply the password, you must supply the username");
	    return false;
	}
	if ((null != username) && (null == password)) {
	    System.err.println("If you supply the username, you must supply the password");
	    return false;
	}

	return true;
    }
    
    private boolean parseArgs(String[] args) {
	if (args.length == 0) {
	    System.err.println("No arguments specified");
	    return false;
	}
	if (0 != args.length % 2)
	    return false;
	Map<String, String> amap = new HashMap<String,String>();
	for (int i = 0; i < args.length; i+=2)
	    amap.put(args[i], args[i+1]);

	host = amap.remove("-host");
	if (null == host) { // host is required
	    System.err.println("Must provide a host");
	    return false;
	}

	filename = amap.remove("-f");
	if (null == filename) { // filename is required
	    System.err.println("Must provide a filename/directory");
	    return false;
	}

	cqlSchema = amap.remove("-schema");
	if (null == cqlSchema) { // schema is required
	    System.err.println("Must provide a schema");
	    return false;
	}

	String tkey;
	if (null != (tkey = amap.remove("-port")))          port = Integer.parseInt(tkey);
	if (null != (tkey = amap.remove("-user")))          username = tkey;
	if (null != (tkey = amap.remove("-pw")))            password = tkey;
	if (null != (tkey = amap.remove("-numFutures")))    numFutures = Integer.parseInt(tkey);
	if (null != (tkey = amap.remove("-maxErrors")))     maxErrors = Integer.parseInt(tkey);
	if (null != (tkey = amap.remove("-skipRows")))      skipRows = Integer.parseInt(tkey);
	if (null != (tkey = amap.remove("-maxRows")))       maxRows = Integer.parseInt(tkey);
	if (null != (tkey = amap.remove("-badDir")))        badDir = tkey;
	if (null != (tkey = amap.remove("-dateFormat")))    dateFormatString = tkey;
	if (null != (tkey = amap.remove("-nullString")))    nullString = tkey;
	if (null != (tkey = amap.remove("-delim")))         delimiter = tkey;
	if (null != (tkey = amap.remove("-delimInQuotes"))) delimiterInQuotes = Boolean.parseBoolean(tkey);
	if (null != (tkey = amap.remove("-decimalDelim"))) {
	    if (tkey.equals(","))
		locale = Locale.FRANCE;
	}
	if (null != (tkey = amap.remove("-boolStyle"))) {
	    boolStyle = BooleanParser.getBoolStyle(tkey);
	    if (null == boolStyle) {
		System.err.println("Bad boolean style.  Options are: " + BooleanParser.getOptions());
		return false;
	    }
	}
	if (null != (tkey = amap.remove("-numThreads")))       numThreads = Integer.parseInt(tkey);
	if (-1 == maxRows)
	    maxRows = Long.MAX_VALUE;
	if (-1 == maxErrors)
	    maxErrors = Long.MAX_VALUE;
	
	if (!amap.isEmpty()) {
	    for (String k : amap.keySet())
		System.err.println("Unrecognized option: " + k);
	    return false;
	}
	return validateArgs();
    }

    private void setup() {
	// Connect to Cassandra
	Cluster.Builder clusterBuilder = Cluster.builder()
	    .addContactPoint(host)
	    .withPort(port)
	    .withLoadBalancingPolicy(new TokenAwarePolicy( new DCAwareRoundRobinPolicy()));
	if (null != username)
	    clusterBuilder = clusterBuilder.withCredentials(username, password);
	cluster = clusterBuilder.build();
	session = cluster.newSession();
    }

    private void cleanup() {
	if (null != session)
	    session.close();
	if (null != cluster)
	    cluster.close();
    }
    
    public boolean run(String[] args) throws IOException, ParseException, InterruptedException, ExecutionException {
	if (false == parseArgs(args)) {
	    System.err.println("Bad arguments");
	    System.err.println(usage());
	    return false;
	}

	// Setup
	setup();
	
	// open file
	BufferedReader reader = null;
	String readerName = "";
	Deque<File> fileList = new ArrayDeque<File>();
	if (filename.equalsIgnoreCase("stdin")) {
	    reader = new BufferedReader(new InputStreamReader(System.in));
	    readerName = "stdin";
	}
	else {
	    File infile = new File(filename);
	    if (infile.isFile()) {
		reader = new BufferedReader(new FileReader(infile));
		readerName = filename;
	    }
	    else {
		File[] inFileList = infile.listFiles();
		if (inFileList.length < 1)
		    throw new IOException("directory is empty");
		for (int i = 0; i < inFileList.length; i++)
		    fileList.push(inFileList[i]);
	    }
	}

	// Launch Threads
	ExecutorService executor;
	long total = 0;
	if (null != reader) {
	    // One file/stdin to process
	    executor = Executors.newSingleThreadExecutor();
	    Callable<Long> worker = new ThreadExecute(cqlSchema, delimiter, 
						      nullString,
						      delimiterInQuotes, 
						      dateFormatString, 
						      boolStyle, locale, 
						      maxErrors, skipRows,
						      maxRows, badDir, reader, 
						      readerName,
						      session, numFutures);
	    Future<Long> res = executor.submit(worker);
	    total = res.get();
	    executor.shutdown();
	}
	else {
	    executor = Executors.newFixedThreadPool(numThreads);
	    Set<Future<Long>> results = new HashSet<Future<Long>>();
	    while (!fileList.isEmpty()) {
		File tFile = fileList.pop();
		BufferedReader r = new BufferedReader(new FileReader(tFile));
		String rName = tFile.getName();
		Callable<Long> worker = new ThreadExecute(cqlSchema, delimiter,
							  nullString,
							  delimiterInQuotes, 
							  dateFormatString, 
							  boolStyle, locale, 
							  maxErrors, skipRows,
							  maxRows, badDir, r, 
							  rName, session,
							  numFutures);
		results.add(executor.submit(worker));
	    }
	    executor.shutdown();
	    for (Future<Long> res : results)
		total += res.get();
	}
	System.err.println("Total rows inserted: " + total);

	// Cleanup
	cleanup();

	return true;
    }

    public static void main(String[] args) throws IOException, ParseException, InterruptedException, ExecutionException {
	CqlDelimLoad cdl = new CqlDelimLoad();
	cdl.run(args);
    }

    class ThreadExecute implements Callable<Long> {
	private Session session;
	private String insert;
	private PreparedStatement statement;
	private CqlDelimParser cdp;
	private long maxErrors;
	private long skipRows;
	private long maxRows;
	private String badDir;
	private BufferedWriter badWriter = null;
	private BufferedReader reader;
	private String readerName;
	private int numFutures;
	private long numInserted;

	private String cqlSchema;
	private Locale locale = null;
	private BooleanParser.BoolStyle boolStyle = null;
	private String nullString = null;
	private String delimiter = null;
	private boolean delimiterInQuotes;

	public ThreadExecute(String inCqlSchema, String inDelimiter, 
			     String inNullString, boolean inDelimiterInQuotes, 
			     String inDateFormatString,
			     BooleanParser.BoolStyle inBoolStyle, 
			     Locale inLocale, 
			     long inMaxErrors, long inSkipRows, long inMaxRows,
			     String inBadDir, BufferedReader inReader,
			     String inReaderName, Session inSession,
			     int inNumFutures) {
	    super();
	    cqlSchema = inCqlSchema;
	    delimiter = inDelimiter;
	    nullString = inNullString;
	    delimiterInQuotes = inDelimiterInQuotes;
	    dateFormatString = inDateFormatString;
	    boolStyle = inBoolStyle;
	    locale = inLocale;
	    maxErrors = inMaxErrors;
	    skipRows = inSkipRows;
	    maxRows = inMaxRows;
	    badDir = inBadDir;
	    reader = inReader;
	    readerName = inReaderName;
	    session = inSession;
	    numFutures = inNumFutures;
	}

	public Long call() throws IOException, ParseException {
	    setup();
	    numInserted = execute();
	    cleanup();
	    return numInserted;
	}

	private void setup() throws IOException, ParseException {
	    // Prepare Badfile
	    if (null != badDir)
		badWriter = new BufferedWriter(new FileWriter(badDir + "/" 
							      + readerName
							      + ".BAD"));
	    
	    cdp = new CqlDelimParser(cqlSchema, delimiter, nullString, 
				     delimiterInQuotes, dateFormatString, 
				     boolStyle, locale);
	    insert = cdp.generateInsert();
	    statement = session.prepare(insert);
	}
	
	private void cleanup() throws IOException {
	    if (null != badWriter)
		badWriter.close();
	}

	private long execute() throws IOException {
	    List<ResultSetFuture> futures = new ArrayList<ResultSetFuture>();
	    String line;
	    int lineNumber = 0;
	    int numInserted = 0;
	    int numErrors = 0;
	    List<Object> elements;

	    System.err.println("*** Processing " + readerName);
	    while ((line = reader.readLine()) != null) {
		lineNumber++;
		if (skipRows > 0) {
		    skipRows--;
		    continue;
		}
		if (maxRows-- < 0)
		    break;
		
		if (0 == line.trim().length())
		    continue;
		
		if (0 == lineNumber % numFutures)
		    purgeFutures(futures);
		
		if (null != (elements = cdp.parse(line))) {
		    BoundStatement bind = statement.bind(elements.toArray());
		    ResultSetFuture resultSetFuture = session.executeAsync(bind);
		    futures.add(resultSetFuture);
		    numInserted++;
		}
		else {
		    System.err.println(String.format("Error parsing line %d in %s: %s", lineNumber, readerName, line));
		    if (null != badWriter) {
			badWriter.write(line);
			badWriter.newLine();
		    }
		    numErrors++;
		    if (maxErrors <= numErrors) {
			System.err.println(String.format("Maximum number of errors exceeded (%d) for %s", numErrors, readerName));
			cluster.close();
			return -1;
		    }
		}
	    }
	    purgeFutures(futures);
	    
	    System.err.println("*** DONE: " + filename + "  number of lines processed: " + lineNumber + " (" + numInserted + " inserted)");

	    return numInserted;
	}

	private void purgeFutures(List<ResultSetFuture> futures) {
	    for (ResultSetFuture future: futures) {
		future.getUninterruptibly();
	    }
	    futures.clear();
	}

	long getNumInserted() {
	    return numInserted;
	}
    }
}

